### DISCUSSION TITLE
product queries / searches in the shop-api don't return promotions or sale prices

### DISCUSSION THREAD

**Discussion by ciocca-dev** [vendurehq/vendure#2977] [General]
Can't product queries / searches in the shop-api return promotions or sale prices?
How is this possible? It's one of the main functions of any e-commerce...

**Comment by michaelbromley** [comment]
Hi, currently this is not supported out of the box. The issue is discussed here:

- https://github.com/vendure-ecommerce/vendure/issues/1365#issuecomment-1017330463

In short, our Promotions system is so dynamic that in order to know the discounted prices, we would need to evaluate all active promotions against every product when displaying it. Since promotions can depend on arbitrary state eg customer group, contents of the active order etc, we cannot cache or pre-compute this in the general case.

Some work arounds:

- If the discount is known to you ahead of time, you could create a custom field to store the discounted price and display this in your storefront. You can also then use this custom field data as the basis of the actual promotion logic.
- You could try to dynamically compute the discounts yourself by defining a custom field resolver on the ProductVariant type, in which you evaluate and apply the active promotions.

**Reply by ciocca-dev** [reply]
This complicates sale management and shifts pricing to the front end. It's not ideal
